state/protocol/snapshot.go

```diff
@@ -71,6 +71,13 @@ type Snapshot interface {
 	// are NOT guaranteed to have been validated by HotStuff.
 	Pending() ([]flow.Identifier, error)
 
+	// ValidPending returns the IDs of all descendants of the Head block.
+	// Valid means the returned bocks have at least one child.
+	// The IDs are ordered such that parents are included before their children. Includes
+	// only blocks that have child blocks.
+	// These are NOT guaranteed to have been validated by HotStuff.
```

Comment:
⚠️ 😨 ⚠️ 
Working with blocks that have not been marked valid _worries me a lot_. Because we generally assume inputs for components to be validated. We are significantly increasing the attack surface  for byzantine nodes, if we process unvalidated blocks. 

Specifically, the existence of a child block does not indicate that the parent is valid in any way.  For example, there could be a long fork of invalid blocks produced by byzantine nodes. They won't have valid QC's ... but byzantine nodes could nevertheless publish such blocks. 

### Suggestion:
I would feel much better, if we only returned blocks that where actually marked as valid: 
* for example, use https://github.com/onflow/flow-go/blob/5433e1d3959767dc5e1f6b3ae28092502e8308b9/storage/badger/operation/headers.go#L35-L36 (and stop traversing a fork as soon as you hit an block _not_ yet marked as valid, because then all its descendants cannot be marked valid either)
* existence of children is in my opinion irrelevant for the context you are using the method `ValidPending`